File: src/main/java/fun/spmc/smpmod/fishing/BiomeCategory.java
package fun.spmc.smpmod.fishing;

import fun.spmc.smpmod.registry.PolymerFishes;
import net.minecraft.core.Holder;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.Identifier;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.tags.BiomeTags;
import net.minecraft.tags.TagKey;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.biome.Biome;

import java.util.*;

public enum BiomeCategory {
    LAVA(PolymerFishes.LAVA),
    VOID(PolymerFishes.END),
    DEEP(PolymerFishes.DEEP),
    SNOWY(PolymerFishes.SNOWY),
    DESERT(PolymerFishes.DESERT),
    TROPICAL(PolymerFishes.TROPICAL),
    PLAINS(PolymerFishes.PLAINS),
    SKY(PolymerFishes.SKY),
    DEFAULT(PolymerFishes.FISH);

    private final List<Item> fishArray;
    private static final Map<Item, BiomeCategory> FISH_TO_CATEGORY = new IdentityHashMap<>();

    BiomeCategory(List<Item> fishArray) { this.fishArray = fishArray; }
    public List<Item> getFishArray() { return fishArray; }

    private static final TagKey<Biome> C_COLD = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_cold"));
    private static final TagKey<Biome> C_SNOWY = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_snowy"));

    private static final TagKey<Biome> C_DESERT = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_desert"));
    private static final TagKey<Biome> C_DRY = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_dry"));

    private static final TagKey<Biome> C_JUNGLE = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_jungle"));
    private static final TagKey<Biome> C_SAVANNA = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_savanna"));
    private static final TagKey<Biome> C_SWAMP = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_swamp"));
    private static final TagKey<Biome> C_TROPICAL = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_tropical"));

    private static final TagKey<Biome> C_OCEAN = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_ocean"));
    private static final TagKey<Biome> C_RIVER = TagKey.create(Registries.BIOME, Identifier.fromNamespaceAndPath("c", "is_river"));

    public static List<Item> getAvailableFish(ServerPlayer player) {
        Holder<Biome> biomeHolder = player.level().getBiome(player.getOnPos());
        double yPos = player.getY();

        Set<BiomeCategory> matchingCategories = EnumSet.noneOf(BiomeCategory.class);

        if (biomeHolder.is(C_SNOWY) || biomeHolder.is(C_COLD)) matchingCategories.add(SNOWY);
        if (biomeHolder.is(C_DESERT) || biomeHolder.is(C_DRY) || biomeHolder.is(BiomeTags.IS_BADLANDS)) matchingCategories.add(DESERT);
        if (biomeHolder.is(C_JUNGLE) || biomeHolder.is(C_SAVANNA) || biomeHolder.is(C_SWAMP) || biomeHolder.is(C_TROPICAL) || biomeHolder.is(BiomeTags.IS_SAVANNA) || biomeHolder.is(BiomeTags.IS_JUNGLE)) matchingCategories.add(TROPICAL);
        if (biomeHolder.is(C_OCEAN) || biomeHolder.is(C_RIVER) || biomeHolder.is(BiomeTags.IS_OCEAN) || biomeHolder.is(BiomeTags.IS_RIVER) || biomeHolder.is(BiomeTags.IS_HILL) || biomeHolder.is(BiomeTags.IS_TAIGA) || biomeHolder.is(BiomeTags.IS_BEACH)) matchingCategories.add(PLAINS);

        if (yPos > 240) {
            matchingCategories.clear();
            matchingCategories.add(SKY);
        }
        if (yPos < 0) {
            matchingCategories.clear();
            matchingCategories.add(DEEP);
        }

        if (player.level().dimension() == ServerLevel.NETHER) {
            matchingCategories.clear();
            matchingCategories.add(LAVA);
        }
        if (player.level().dimension() == ServerLevel.END) {
            matchingCategories.clear();
            matchingCategories.add(VOID);
        }

        List<Item> combinedFish = new ArrayList<>();
        for (BiomeCategory category : matchingCategories) combinedFish.addAll(category.getFishArray());
        combinedFish.addAll(DEFAULT.getFishArray());
        return combinedFish;
    }

    public static void initLookupMap() { // init
        FISH_TO_CATEGORY.clear();
        for (BiomeCategory category : values()) {
            if (category == DEFAULT) continue;
            for (Item fish : category.getFishArray()) FISH_TO_CATEGORY.putIfAbsent(fish, category);
        }
    }

    public String toString() { return name().charAt(0) + name().substring(1).toLowerCase(); }

    public static BiomeCategory getBiomeFish(Item fish) {
        return FISH_TO_CATEGORY.getOrDefault(fish, DEFAULT);
    }
}

File: src/main/java/spmc/smpmod/core/BiomeCategory.kt
package spmc.smpmod.core

import net.fabricmc.fabric.api.tag.convention.v2.ConventionalBiomeTags
import net.minecraft.core.Holder
import net.minecraft.resources.ResourceKey
import net.minecraft.tags.BiomeTags
import net.minecraft.tags.TagKey
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.biome.Biome
import spmc.smpmod.utils.UtilFunc.createBiomeTag
import java.util.*
import java.util.concurrent.ConcurrentHashMap

enum class BiomeCategory(val tags: List<TagKey<Biome>> = emptyList()) {
	BADLANDS(listOf(BiomeTags.IS_BADLANDS)),
	DESERT(listOf(ConventionalBiomeTags.IS_DRY_OVERWORLD, ConventionalBiomeTags.IS_DESERT)),
	SNOWY(listOf(ConventionalBiomeTags.IS_SNOWY, ConventionalBiomeTags.IS_COLD_OVERWORLD)),
	TROPICAL(listOf(ConventionalBiomeTags.IS_JUNGLE, ConventionalBiomeTags.IS_SAVANNA, ConventionalBiomeTags.IS_SWAMP, createBiomeTag("is_tropical"), BiomeTags.IS_JUNGLE, BiomeTags.IS_SAVANNA)), // TODO: remove tropical tag with ConventionalBiomeTags
	OCEAN(listOf(ConventionalBiomeTags.IS_OCEAN, ConventionalBiomeTags.IS_RIVER, BiomeTags.IS_OCEAN, BiomeTags.IS_RIVER, BiomeTags.IS_BEACH)),
	FOREST(listOf(ConventionalBiomeTags.IS_FOREST)), // TODO
	TAIGA(listOf(BiomeTags.IS_TAIGA, ConventionalBiomeTags.IS_TAIGA)),
	PLAINS(listOf(BiomeTags.IS_HILL, ConventionalBiomeTags.IS_HILL)), // TODO: check
	FLOWER(listOf(ConventionalBiomeTags.IS_FLORAL, ConventionalBiomeTags.IS_FLOWER_FOREST)), // TODO
	MOUNTAIN(listOf(BiomeTags.IS_MOUNTAIN)), // TODO: checks "peaks", "slopes", "stony", "windswept"
	MUSHROOM(listOf(BiomeTags.WITHOUT_ZOMBIE_SIEGES)),

	CAVE(listOf(ConventionalBiomeTags.IS_CAVE)),
	SCULK(listOf(BiomeTags.HAS_ANCIENT_CITY)), // TODO

	DEEP,
	SKY,

	NETHER(listOf(BiomeTags.IS_NETHER)), // general nether
	CRIMSON,
	WARPED,
	SOUL_SAND,
	BASALT,
	NETHER_WASTES,

	END(listOf(BiomeTags.IS_END)), // probably includes every end biome? TODO: add nullscape diversity (not a good idea to put here)
	// END_ISLANDS? Identifier.fromNamespaceAndPath("c", "end_islands"))
	DEFAULT;

	override fun toString() = name[0].toString() + name.substring(1).lowercase(Locale.getDefault())
	companion object {
		val OVERWORLD = listOf(BADLANDS, DESERT, CAVE, FOREST, SCULK, MUSHROOM, TROPICAL, TAIGA, OCEAN, FLOWER, MOUNTAIN, SNOWY, SKY)

		private val BIOME_CACHE = ConcurrentHashMap<ResourceKey<Biome>, BiomeCategory>()

		fun getGroup(biome: Holder<Biome>): BiomeCategory = BIOME_CACHE.computeIfAbsent(biome.unwrapKey().orElse(null) ?: return DEFAULT) { BiomeCategory.entries.firstOrNull { group -> group != DEFAULT && (group.tags.any { biome.`is`(it) }) } ?: DEFAULT } // TODO: EITHER remove edge cases / change how it works (NOPE, adding own biome tags)
		fun getPlayerCategories(player: Player): Set<BiomeCategory> {
			val level = player.level()
			val yPos = player.y

			if (level.dimension() == Level.NETHER) {
				val biome = getGroup(level.getBiome(player.blockPosition()))
				return if (biome != DEFAULT) EnumSet.of(biome) else EnumSet.of(NETHER_WASTES)
			}
			if (level.dimension() == Level.END) return EnumSet.of(END)

			if (yPos > 240) return EnumSet.of(SKY)
			if (yPos < 0) return EnumSet.of(DEEP) // TODO: merge sculk with deep? / vice versa? (only for fishing)

			val biomeHolder = level.getBiome(player.blockPosition())
			val category = getGroup(biomeHolder)
			return if (category == DEFAULT) EnumSet.of(DEFAULT) else EnumSet.of(category)
		}
	}
}